internal/backend/utils.go

```diff
@@ -20,6 +23,13 @@ func LoadAll(ctx context.Context, buf []byte, be restic.Backend, h restic.Handle
 			return cerr
 		}
 		buf = wr.Bytes()
+
+		id, err := restic.ParseID(h.String())
+		if !retriedInvalidData && err == nil && h.Type != restic.ConfigFile && !restic.Hash(buf).Equal(id) {
```

Comment:
This is the line I mean: I have a hard time understanding what's tested here. Hm.

Reply:
I've slightly restructured the check. The outer check now tests that this is the first try and it's not the config file.
Then if the ID parsing is successful, then it checks whether the hash is wrong. In that case it returns an error that either will cause the cache to redownload the file or just a plain retry.